flang/lib/Lower/IO.cpp

```diff
@@ -729,6 +729,15 @@ static const Fortran::semantics::SomeExpr *getExpr(const A &stmt) {
   llvm_unreachable("must have a file unit");
 }
 
+/// Get the expression from a list for non-wrapper classes.
+template <typename SEEK, typename A>
+static const Fortran::semantics::SomeExpr *getExprInList(const A &stmt) {
```

Comment:
It seems this search should only be done once.

INQUIRE will either have a FileUnitNumber or a FileNameExpr, and never both. The algorithm can be simplified to scan the list, find one or the other, and return the expression and an indication of which type was found.